### PR TITLE
Autocomplete: add `autocomplete.languages` user setting

### PR DESCRIPTION
**Pull request by valerybugakov**
## Context

- Closes https://github.com/sourcegraph/cody/issues/1289
- Adds `cody.autocomplete.languages` to `Enable or disable inline code suggestions for specified languages`

## Screenshots

Screenshot 2023-10-03 at 11 45 04

Screenshot 2023-10-03 at 11 45 14

## Test plan

Tested E2E locally + added unit tests.

### CODE REVIEW

File: vscode/src/completions/create-inline-completion-item-provider.ts [hunk 1]
```diff
@@ -111,3 +114,35 @@ export async function createInlineCompletionItemProvider({
...
+
+        return languageIds
+            .map(language => {
+                if (perLanguageConfig[language] === undefined || perLanguageConfig[language] === true) {
+                    return { language, scheme: 'file' }
+                }
+
+                return null
+            })
+            .filter(isDefined)
```
Comment (dominiccooney) [inline vscode/src/completions/create-inline-completion-item-provider.ts:135]: You can do this more simply with `flatMap`
```suggestion
            .flatMap(language => perLanguageConfig[language] === undefined || perLanguageConfig[language] === true ? [{ language, scheme: 'file' }] : [])
```

File: vscode/src/completions/create-inline-completion-item-provider.ts [hunk 2]
```diff
@@ -111,3 +114,35 @@ export async function createInlineCompletionItemProvider({
...
+    // Enable only for explicitly enabled languages in `perLanguageConfig`.
+    return Object.entries(perLanguageConfig)
+        .map(([language, isEnabled]) => {
+            if (isEnabled) {
+                return { language, scheme: 'file' }
+            }
+
+            return null
+        })
+        .filter(isDefined)
```
Comment (dominiccooney) [inline vscode/src/completions/create-inline-completion-item-provider.ts:147]: Ditto here.
```suggestion
    return Object.entries(perLanguageConfig)
        .flatMap(([language, isEnabled]) => isEnabled ? [{ language, scheme: 'file' }] : [])
```

File: vscode/src/completions/create-inline-completion-item-provider.ts [hunk 3]
```diff
@@ -111,3 +114,35 @@ export async function createInlineCompletionItemProvider({
...
+    autocompleteLanguages: Record<string, boolean>
+): Promise<vscode.DocumentFilter[]> {
+    const { '*': isEnabledForAll, ...perLanguageConfig } = autocompleteLanguages
+
+    // Enable for every known langauge ID if it's not explicitly disabled in `perLanguageConfig`.
+    if (isEnabledForAll) {
+        const languageIds = await vscode.languages.getLanguages()
+
+        return languageIds
+            .map(language => {
```
Comment (dominiccooney) [inline vscode/src/completions/create-inline-completion-item-provider.ts:128]: Let's try to de-dup these three similar loops.

File: vscode/src/testutils/mocks.ts [hunk 4]
```diff
@@ -514,6 +514,108 @@ const workspaceFs: Partial<vscode_types.FileSystem> = {
     },
 }
 
+const languages: Partial<typeof vscode_types.languages> = {
+    getLanguages() {
```
Comment (dominiccooney) [inline vscode/src/testutils/mocks.ts:518]: Would be great to drop a pointer to where this is upstream so it is easy to keep up to date.

**Review by dominiccooney — APPROVED** [review]
Looks good but `flatMap` returning [] to skip items, and less duplication of those loops, would be nice.

**Review by dominiccooney — APPROVED** [review]
It reads like poetry. Great work.

**Review by umpox — APPROVED** [review]

**Review by philipp-spiess — APPROVED** [review]

**Pull request merged**